File: marguerite/marguerite.erl
-module(marguerite).
-export([jwt_sign/6, jwt_verify/3, aes_encrypt/1, aes_decrypt/1, hmac_sign/1, hmac_verify/2, version/0]).
-nifs([jwt_sign/6, jwt_verify/3, aes_encrypt/1, aes_decrypt/1, hmac_sign/1, hmac_verify/2, version/0]).
-on_load(init/0).

init() ->
    ok = erlang:load_nif("./libmarguerite", 0).

jwt_sign(_U, _V, _W, _X, _Y, _Z) ->
    erlang:nif_error(nif_library_not_loaded).

jwt_verify(_X, _Y, _Z) ->
    erlang:nif_error(nif_library_not_loaded).

aes_encrypt(_X) ->
    erlang:nif_error(nif_library_not_loaded).

aes_decrypt(_X) ->
    erlang:nif_error(nif_library_not_loaded).

hmac_sign(_X) ->
    erlang:nif_error(nif_library_not_loaded).

hmac_verify(_X, _Y) ->
    erlang:nif_error(nif_library_not_loaded).

version() ->
    erlang:nif_error(nif_library_not_loaded).

File: jasmine/lib/marguerite.ex
defmodule Marguerite.NIF do
  @on_load :init
  def init do
    path = :filename.join([:code.priv_dir(:jasmine), "libmarguerite"])
    :erlang.load_nif(path, 0)
  end

  def jwt_sign(_v, _w, _x, _y, _z), do: :erlang.nif_error(:nif_library_not_loaded)
  def jwt_sign(_u, _v, _w, _x, _y, _z), do: :erlang.nif_error(:nif_library_not_loaded)
  def jwt_verify(_x, _y, _z), do: :erlang.nif_error(:nif_library_not_loaded)

  def hmac_sign(_x), do: :erlang.nif_error(:nif_library_not_loaded)
  def hmac_verify(_x, _y), do: :erlang.nif_error(:nif_library_not_loaded)

  def aes_encrypt(_x), do: :erlang.nif_error(:nif_library_not_loaded)
  def aes_decrypt(_x), do: :erlang.nif_error(:nif_library_not_loaded)

  def s3_create_bucket(_x), do: :erlang.nif_error(:nif_library_not_loaded)
  def s3_create_bucket(_x, _y), do: :erlang.nif_error(:nif_library_not_loaded)
  def s3_create_bucket(_x, _y, _z), do: :erlang.nif_error(:nif_library_not_loaded)
  def s3_put_object(_x, _y), do: :erlang.nif_error(:nif_library_not_loaded)
  def s3_get_presigned_object_url(_x, _y), do: :erlang.nif_error(:nif_library_not_loaded)
  def s3_get_presigned_object_url(_x, _y, _z), do: :erlang.nif_error(:nif_library_not_loaded)
  def s3_get_permanent_object_url(_x, _y), do: :erlang.nif_error(:nif_library_not_loaded)

  def version(), do: :erlang.nif_error(:nif_library_not_loaded)
end
